sdk/webpubsub/azure-messaging-webpubsub-client/src/test/java/com/azure/messaging/webpubsub/client/ClientTests.java

```diff
@@ -0,0 +1,451 @@
+        Assertions.assertEquals("message1", messageReceived.iterator().next());
+    }
+
+    @Test
+    @DoNotRecord(skipInPlayback = true)
+    public void testConnectedDisconnectedEvent() throws InterruptedException {
+        CountDownLatch latch = new CountDownLatch(1);
+        List<String> eventReceived = new ArrayList<>();
+
+        WebPubSubClient client = getClientBuilder()
+            .buildClient();
+
+        client.addOnConnectedEventHandler(event -> eventReceived.add(event.getClass().getSimpleName()));
+        client.addOnDisconnectedEventHandler(event -> eventReceived.add(event.getClass().getSimpleName()));
+        client.addOnStoppedEventHandler(event -> latch.countDown());
+
+        client.start();
+        client.stop();
+
+        latch.await(1, TimeUnit.SECONDS);
+
+        Assertions.assertEquals(2, eventReceived.size());
+        Assertions.assertEquals(ConnectedEvent.class.getSimpleName(), eventReceived.get(0));
+        Assertions.assertEquals(DisconnectedEvent.class.getSimpleName(), eventReceived.get(1));
+    }
+
+    @Test
+    @DoNotRecord(skipInPlayback = true)
+    public void testServerDisconnectedOnInvalidPayload() throws InterruptedException {
+        CountDownLatch latch = new CountDownLatch(1);
+        List<String> eventReceived = new ArrayList<>();
+
+        WebPubSubClient client = getClientBuilder()
+            .buildClient();
+
+        client.addOnConnectedEventHandler(event -> eventReceived.add(event.getClass().getSimpleName()));
+        client.addOnDisconnectedEventHandler(event -> eventReceived.add(event.getClass().getSimpleName()));
+        client.addOnStoppedEventHandler(event -> latch.countDown());
+
+        client.start();
+
+        client.getWebsocketSession().sendTextAsync("invalid", result -> {
+            // NOOP
+        });
+
+        // wait for 1008
+
+        latch.await(5, TimeUnit.SECONDS);
+        Assertions.assertEquals(0, latch.getCount());
+
+        Assertions.assertEquals(2, eventReceived.size());
+        Assertions.assertEquals(ConnectedEvent.class.getSimpleName(), eventReceived.get(0));
+        Assertions.assertEquals(DisconnectedEvent.class.getSimpleName(), eventReceived.get(1));
+    }
+
+    @Test
+    @DoNotRecord(skipInPlayback = true)
+    public void testServerDisconnectedEventOnSocketClose() throws InterruptedException {
+        CountDownLatch latch = new CountDownLatch(1);
+        List<String> eventReceived = new ArrayList<>();
+
+        WebPubSubClient client = getClientBuilder()
+            .buildClient();
+
+        client.addOnConnectedEventHandler(event -> eventReceived.add(event.getClass().getSimpleName()));
+        client.addOnDisconnectedEventHandler(event -> eventReceived.add(event.getClass().getSimpleName()));
+        client.addOnStoppedEventHandler(event -> latch.countDown());
+
+        client.start();
+
+        client.getWebsocketSession().closeSocket();
+
+        // client would recover
+        Thread.sleep(100);
```

Comment:
Here we want to simulate connect event?

Reply:
It calls close socket `client.getWebsocketSession().closeSocket();` that force close the connection. That triggers recover.

Reply:
I am trying to understand the two tests: this one and `testConnectedDisconnectedEvent()`.
Both are validating `ConnectedEvent` and `DisconnectedEvent`. 
In `testConnectedDisconnectedEvent()`, it triggers `ConnectedEvent` and `DisconnectedEvent` by `client.start` and `client.stop`, then validate the two events are triggered once. In this test, aprart from `client.start` and `client.stop`, we also close socket and wait for it to recover, if start and stop client will trigger  `ConnectedEvent` and `DisconnectedEvent` events, what do we want to validate by adding socket close and recovery, will it trigger events?

Reply:
recover does not count as ConnectedEvent or DisconnectedEvent. If recover success, it is same connection.

Reply:
I may not have a good name on the test.

It is not exactly "testServerDisconnectedEventOnSocketClose", because server does not send "DisconnectedEvent".

Maybe just "testServerRecoverOnSocketClose"

Reply:
I don't quite get the validation part of the case, if we want to validate the recovery process on socket close, no matter recovery succeeds or fails, seems it will not change received events?

Reply:
It is very hard to test on the "recovery", as it is internal state and transient.

Maybe we can only test, that after "recovery", the connection still works, maybe by adding another send and verify it works.

Reply:
Yeah, it would be better and more straightforward to test the connection is still alive.

Reply:
Despite this one, other tests and codes look good to me now.